File: Partition Array Such That Maximum Difference Is K/Partition Array Such That Maximum Difference Is K.cpp
// Solution for Partition Array Such That Maximum Difference Is K in CPP

class Solution {
public:
    int partitionArray(vector<int>& nums, int k) {
        // Sort the array to group close elements together
        sort(nums.begin(), nums.end());

        int minSubsequences = 1;           // At least one subsequence is needed
        int minInSubsequence = nums[0];    // Minimum element of the current subsequence

        for (const int& ele: nums) {
            // If the current number cannot fit in the current subsequence
            if (ele - minInSubsequence > k) {
                minSubsequences++;         // Start a new subsequence
                minInSubsequence = ele; // Update minimum of the new subsequence
            }
        }

        return minSubsequences;
    }
};

File: Partition Array Such That Maximum Difference Is K/Partition Array Such That Maximum Difference Is K.py
# Solution for Partition Array Such That Maximum Difference Is K in PY

class Solution:
    def partitionArray(self, nums: List[int], k: int) -> int:
        nums.sort()

        minSubsequences = 1
        minInSubsequence = nums[0]

        for ele in nums:
            if ele - minInSubsequence > k:
                minSubsequences += 1
                minInSubsequence = ele

        return minSubsequences
